### PR TITLE
Refactor parse' to use tail recursion

### PR DESCRIPTION
**Pull request by markfields**
This should bring no change in behavior, only a resiliency against `StackOverflowException` (see #1277).

There's one non-tail-recursive call in `parse'`, which is the one when stepping into a tag
to parse its children. This change adds a manual callstack to the function
onto which we push the variables in the closure before continuing to the
child tokens. Then in the places where we were returning, call a helper
to pop the stack and include the logic that originally followed the recursive call.

All other recursive calls are already tail recursive, and represent a go-to-next-sibling step rather than a go-to-first-child step in the resulting HTML.  So they remain unchanged.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -23,4 +23,5 @@ paket-files
 TestResults
 *.log
 .vscode
-.idea
\ No newline at end of file
+.idea
+.ionide/
```
Comment (markfields) [inline .gitignore]: I can revert this change if requested; it was necessary in my dev environment where I use the ionide plugin for VS Code.

**Comment by markfields** [comment]
@ovatsus care to take a look at this change?

I haven't been able to sort out the Travis failure, and I'm also unable to build/run the tests locally, so I wouldn't merge until I get that figured out.  But looking for feedback on the concept/implementation.  Thanks.

**Comment by dsyme** [comment]
Reopen for travis

**Comment by markfields** [comment]
Wow forgot about this. Not sure what to make of all the timeout errors in the AppVeyor test run. @dsyme do you suppose they're related to this change or not? Skimming the callstacks I don't see anything that obviously would be doing HTML Parsing. Obviously if I wrote an infinite loop that would cause timeouts though... :)

**Comment by cartermp** [comment]
@markfields sorry for no response here - we've reworked several aspects of this repository and I'm scanning through contributions that can be taken in now. This one looks good, subject to tests passing. No more AppVeyor :)

**Comment by markfields** [comment]
Thanks @cartermp!

**Pull request merged**